pylint/extensions/code_style.py

```diff
@@ -135,6 +152,79 @@ def _check_dict_consider_namedtuple_dataclass(self, node: nodes.Dict) -> None:
             self.add_message("consider-using-namedtuple-or-dataclass", node=node)
             return
 
+    def _check_consider_using_assignment_expr(self, node: nodes.If) -> None:
+        """Check if an assignment expression (walrus operator) can be used.
+
+        For example if an assignment is directly followed by an if statment:
+        >>> x = 2
+        >>> if x:
+        >>>     ...
+
+        Can be replaced by:
+        >>> if (x := 2):
+        >>>     ...
+
+        Note: Assignment expressions were added in Python 3.8
+        """
+        node_name: Optional[nodes.Name] = None
+        if isinstance(node.test, nodes.Name):
+            node_name = node.test
+        elif (
+            isinstance(node.test, nodes.UnaryOp)
+            and node.test.op == "not"
+            and isinstance(node.test.operand, nodes.Name)
+        ):
+            node_name = node.test.operand
+        elif (
+            isinstance(node.test, nodes.Compare)
+            and isinstance(node.test.left, nodes.Name)
+            and len(node.test.ops) == 1
+        ):
+            node_name = node.test.left
+        else:
+            return
+
+        prev_sibling = node.previous_sibling()
+        if self._check_prev_sibling_to_if_stmt(prev_sibling, node_name.name):
+            test_str = node.test.as_string().replace(
+                node_name.name,
+                f"({node_name.name} := {prev_sibling.value.as_string()})",  # type: ignore
+                1,
+            )
+            suggestion = f"if {test_str}:"
+            if len(suggestion) > 88:
```

Comment:
I haven't pushed that latest changes yet as I'm still working on it. There I added a new config value so users can change it.

> If we really want to do this I imagine we should behave differently if line-too-long is disabled or not and we should review all the checkers to see if it would also make sense...

I don't think that is necessary. Personally, I consider assignment expressions a special case as they are a truly optional feature that should only be used where it makes sense. It is completely independent from `line-too-long`.

Furthermore, I can't thing of any specific existing and none optional check that could benefit from such a setting. At least not of the top of my head. Once we add it and users would expect it to work for check x / y that doesn't work yet, I'm sure we would see requests for it and can change it then.

> But beside that I think it's counterintuitive to have a message change if the line is shorter or longer. I don't think we want to surprise user like that. [...] I think it's more reasonable to let the user know all the time that they can use a walrus operator, and if the line become too long because of it, they could refactor it so it's shorter.

Again, I would disagree here. A refactor is not always possible (or easy do accomplish). It's much more helpful to only recommend it in cases where there is a clear benefit from using it. It's just a recommendation for an optional feature after all.

One example, I think users won't like it if we recommend the use of `:=`, but it results in an additional line because black reformatted it. Now they will need to add a `pylint: disable`. If they do it two, three times, they will just disable the check completely and don't get anything from it. That doesn't help anyone.

Reply:
> It is completely independent from line-too-long.

I don't think a checker that change behavior based on line length should be separated from our max-line-length option. If the value is 120 in max-line-length it means the user want to make 120 char lines. It would be strange to not raise a warning because more than 88 char is considered too much by this checker, when the user already explicitly said that 120 is okay.